documentation/projects/proposals/dark_mode/20240325-implementation_plan_dark_mode.md

```diff
@@ -0,0 +1,263 @@
+completed in parallel.
+
+1. **Work Stream A**: Implement the new color palette.
+
+   1. Create a `FORCE_DARK_MODE` feature flag which is "off" by default and
+      "switchable" in our staging environment. Add a `dark-mode` class to the
+      root HTML tag when this flag is enabled, and a `light-mode` class which is
+      set by default. This will not result in any visual changes.
+   2. The following steps can take place in parallel:
+
+      1. Rename all colors in the Tailwind configuration and the frontend
+         components to use new, semantic names (specific names TBD). This will
+         not result in any visual changes. **This is likely to be the largest PR
+         to review as it is a global find/replace across the entire frontend.**
+
+         1. Replace the "hardcoded" color values in the Tailwind configuration
+            file with css variables defined in the "base" layer of the
+            `tailwind.css` file. This will not result in any visual changes.
+
+      2. Add the dark mode colors as CSS variable definitions, nested under the
+         `.dark-mode` CSS class, in the "base" layer of the `tailwind.css` file.
+         This will not result in any visual changes, _except when
+         `FORCE_DARK_MODE` is enabled._
+
+      3. Visual Regression tests. Update
+         `frontend/test/playwright/utils/breakpoints.ts` so that each breakpoint
+         produces and expects a dark mode screenshot to pass as well as the
+         existing light mode screenshot. **This will also be a significant diff,
+         as at the time of writing it will create 293 new screenshots to review.
+         This is also the point of the process where @fcoveram and
+         @wordpress/openverse-frontend should review the full dark mode
+         appearance for correctness and sufficient color contrast (see the
+         ["Accessibility"](#accessibility) section for more details.**
+
+2. **Work Stream B**: Toggling dark mode
+
+   1. Create a `DARK_MODE_UI_TOGGLE` feature flag which is off by default and
+      switchable in staging.
+   2. Implement logic for calculating a current "color mode" with the following
+      state:
+
+      ```ts
+      interface ColorMode {
+        preference: "dark" | "light" | "system" // Defaults to "light"
+        systemValue: "light" | "dark" // Readonly representation of the system value
+      }
+      ```
+
+      The color mode should be stored in a cookie so that a previously-selected
+      user choice can be used when rendering via SSR and prevent a visual flash
+      of light mode styles for users who have selected dark mode. The system
+      value is read by matching the `'(prefers-color-scheme: dark)'` media
+      query.
+
+   3. Behind the feature flag, add the new user interface element which toggles
+      dark mode (exact design TBD, but it will be comprised of existing UI
+      components). Default to "light" mode but support choosing between "dark",
+      "light", and "system".
```

Comment:
Shouldn't it default to `system` for user's with `prefers-color-scheme: dark` set? Or does that come later?

Reply:
This is a bit of an open question IMO. Originally @fcoveram wanted to default to light mode so that the site wouldn't suddenly change to dark mode for users, but the more I think about it defaulting to system seems preferable and like a very clean way to _introduce_ the feature to the very users most likely to use it. This is also being discussed a bit in #4157.

Reply:
> …defaulting to system seems preferable and like a very clean way to introduce the feature to the very users most likely to use it…

I revisited this option with other designer contributors, and it seems the best one. No matter the design we choose in #4157, we can safely go with this option.

Reply:
We are now defaulting to system.